### ISSUE TITLE
bam.fetch not working properly

### ISSUE THREAD

**Issue by najoshi** [betteridiot/bamnostic#49]
I am trying to fetch alignments from a bam file. There are definitely alignments between 10000000 and 60000000 on CHR1. That fetch works. But when I do the exact same fetch using 11000000 and 60000000 on CHR1, I get nothing. And there are definitely alignments between 11000000 and 60000000.

This has output (and alignments well into 50Mbp):

    for read in bam.fetch('CHR1', 10000000, 60000000):
        print(read)

This does not print anything:

    for read in bam.fetch('CHR1', 11000000, 60000000):
        print(read)

 - OS: Ubuntu 18.04
 - Python Version: 3.6.9
 - bamnostic Version: 1.1.8

[align.bam.bai.gz](https://github.com/betteridiot/bamnostic/files/8123861/align.bam.bai.gz)
[align.bam.gz](https://github.com/betteridiot/bamnostic/files/8123862/align.bam.gz)

**Comment by betteridiot** [comment]
Alright @najoshi. I have been working on this nonstop trying to figure out what was wrong with `fetch`. When I thought I had it solved, turned out I was wrong. 

**TL;DR**: your `align.bam.bai` file was the culprit. After I re-indexed your file with samtools, I didn't have any problems anymore. Something with the original index file was corrupted and was assigning incorrect linear indices to specific bins.

---
# Bug Investigation
## Environment:
I created a conda environment for testing on my Windows Subsystem for Linux running Ubuntu 18.04.4 LTS
`conda create -n bamnostic_issue python samtools bamnostic pysam`
I also installed some basic QoL tools that allowed me to investigate using Jupyter Lab.

## Issue Verification
I wanted to make sure that I could replicate your issue. Below are the steps 
![image](https://user-images.githubusercontent.com/25447319/164771632-b4b5f6bc-eeb6-4fd8-9cbd-76fac1fcd16e.png)

First, I wanted to see if I could accomplish the same output as `pysam` for the "good" region you provided:
![image](https://user-images.githubusercontent.com/25447319/164771722-e51a2b4c-f6d4-4067-8a30-2b1a5be7ed43.png)

Then I wanted to replicate the issue you brought up with the "bad" region:
![image](https://user-images.githubusercontent.com/25447319/164771810-62425ab2-8f47-4923-af38-1bcf9069648a.png)

This is where I saw what you were mentioning. So I started tracing the issue.  Internally, `bamnostic.bam.fetch` was returning a None on line 593, which implies no reads being available. This is because `bamnostic.bai.query` was not returning a valid read block. Looking at the "good" region for the index query showed that the first `linear_offset` was correctly associated with the first read block bin offsets (first line of the output below). More importantly is to take notice of the second line of the output 85066353 represents the _next_ bins offset:
![image](https://user-images.githubusercontent.com/25447319/164772477-43951dc7-e1ea-4c52-b568-86005b1e7048.png)

However, when you look at the "bad" region (below), that the first bin's offset (85066353) is 1 ahead of where it is supposed to be (85066170). This meant that the query function started "looking" for the correct bin too late. That is, it was like looking up the definition of "Apple" in the dictionary, but starting at "Bark" instead.
![image](https://user-images.githubusercontent.com/25447319/164772725-b823a493-4d92-4262-8300-2239b4accacf.png)
Investigating the predicted bins of `reg2bins` for both the "good" and "bad" regions showed this prediction.
![image](https://user-images.githubusercontent.com/25447319/164773526-63ceedfd-7e25-4aca-bae3-c61d6f324e90.png)
Notice how the "good" regions goes from bin 1042 to 5291 (the correct bin), but the "bad" region goes from 1042 to 5352. 

This got me thinking that the `bamnostic.bai.reg2bins` function was improperly computing bins. The function is a Python port of the original `reg2bins` function listed in the HTS-spec: https://samtools.github.io/hts-specs/SAMv1.pdf. I rewrote a version of `reg2bins` to see what is happening with the bitwise shifts and further noticed that it starts further upstream for the "bad" region compared to the "good" region.
![image](https://user-images.githubusercontent.com/25447319/164773875-e408317e-5792-40ef-97e3-4d43adc34e36.png)
The shift on lines 5 and offset on line 6 indicate what I thought was an "off by 1" error somewhere. For some reason, the index was labeling multiple bins as having the same liner offest
![image](https://user-images.githubusercontent.com/25447319/164774807-3130d495-9052-4f29-a27c-f0460256868a.png)
Notice that number? 85066170. We have seen that before.

That led me to test one more thing: _the input data_. Devs and IT say it all the time: "have you tried turning it off and turning it back on again?" I had assumed that it was my code that was wrong. But I wanted to test to make sure.

## The final check
I used `samtools` to index your original bam file to make a new `align.bam.bai` file.
![image](https://user-images.githubusercontent.com/25447319/164775574-3336c78d-3bef-42d5-b9b1-4a018c93af7d.png)
After making a new index, I ran the original code again against both the "good" and "bad" regions compared to `pysam`'s output and found I was getting the correct output.
![image](https://user-images.githubusercontent.com/25447319/164775855-17278809-c0cf-408c-9e6e-044a9ddba226.png)
![image](https://user-images.githubusercontent.com/25447319/164775888-6a25eede-9879-4281-84d4-c6e1987ee038.png)

## The Verdict
I cannot explain why the original output didn't match `pysam`'s (and, by extension `htslib.c`). I suspect that the C-level query function is more robust to allow for these issues. I do suspect that given some version changes in the BAM spec and `samtools` may have led to some instability or a partially corrupted index.

Nevertheless, if you reindex your file, you should be able to achieve the results you were looking for. Sorry again about the delay in response, but I hope this helps. Thank you for your bug report.
